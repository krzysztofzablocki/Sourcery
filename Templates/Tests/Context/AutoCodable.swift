//
//  AutoCodable.swift
//  TemplatesTests
//
//  Created by Ilya Puchka on 13/04/2018.
//  Copyright Â© 2018 Pixle. All rights reserved.
//

import Foundation

protocol AutoDecodable: Swift.Decodable {}
protocol AutoEncodable: Swift.Encodable {}
protocol AutoCodable: AutoDecodable, AutoEncodable {}

public struct CustomKeyDecodable: AutoDecodable {
    let stringValue: String
    let boolValue: Bool
    let intValue: Int

    enum CodingKeys: String, CodingKey {
        case intValue = "integer"

// sourcery:inline:auto:CustomKeyDecodable.CodingKeys.AutoCodable
        case stringValue
        case boolValue
// sourcery:end
    }

}

public struct CustomMethodsCodable: AutoCodable {
    let boolValue: Bool
    let intValue: Int?
    let optionalString: String?
    let requiredString: String
    let requiredStringWithDefault: String

    var computedPropertyToEncode: Int {
        return 0
    }

    static let defaultIntValue: Int = 0
    static let defaultRequiredStringWithDefault: String = ""

    static func decodeIntValue(from container: KeyedDecodingContainer<CodingKeys>) -> Int? {
        return (try? container.decode(String.self, forKey: .intValue)).flatMap(Int.init)
    }

    static func decodeBoolValue(from decoder: Decoder) throws -> Bool {
        return try decoder.container(keyedBy: CodingKeys.self).decode(Bool.self, forKey: .boolValue)
    }

    func encodeIntValue(to container: inout KeyedEncodingContainer<CodingKeys>) {
        try? container.encode(String(intValue ?? 0), forKey: .intValue)
    }

    func encodeBoolValue(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(boolValue, forKey: .boolValue)
    }

    func encodeComputedPropertyToEncode(to container: inout KeyedEncodingContainer<CodingKeys>) {
        try? container.encode(computedPropertyToEncode, forKey: .computedPropertyToEncode)
    }

    func encodeAdditionalValues(to encoder: Encoder) throws {

    }

}

public struct CustomContainerCodable: AutoCodable {
    let value: Int

    enum CodingKeys: String, CodingKey {
        case nested
        case value
    }

    static func decodingContainer(_ decoder: Decoder) throws -> KeyedDecodingContainer<CodingKeys> {
        return try decoder.container(keyedBy: CodingKeys.self)
            .nestedContainer(keyedBy: CodingKeys.self, forKey: .nested)
    }

    func encodingContainer(_ encoder: Encoder) -> KeyedEncodingContainer<CodingKeys> {
        var container = encoder.container(keyedBy: CodingKeys.self)
        return container.nestedContainer(keyedBy: CodingKeys.self, forKey: .nested)
    }
}

struct CustomCodingWithNotAllDefinedKeys: AutoCodable {
    let value: Int
    var computedValue: Int { return 0 }

    enum CodingKeys: String, CodingKey {
        case value

// sourcery:inline:auto:CustomCodingWithNotAllDefinedKeys.CodingKeys.AutoCodable
        case computedValue
// sourcery:end
    }

    func encodeComputedValue(to container: inout KeyedEncodingContainer<CodingKeys>) {
        try? container.encode(computedValue, forKey: .computedValue)
    }

}

struct SkipDecodingWithDefaultValueOrComputedProperty: AutoCodable {
    let value: Int
    let skipValue: Int = 0
    var computedValue: Int { return 0 }

    enum CodingKeys: String, CodingKey {
        case value
        case computedValue
    }
}

struct SkipEncodingKeys: AutoCodable {
    let value: Int
    let skipValue: Int

    enum SkipEncodingKeys {
        case skipValue
    }
}

enum SimpleEnum: AutoCodable {
    case someCase
    case anotherCase
}

enum AssociatedValuesEnum: AutoCodable, Equatable {
    case someCase(id: Int, name: String)
    case anotherCase

    enum CodingKeys: String, CodingKey {
        case enumCaseKey = "type"

// sourcery:inline:auto:AssociatedValuesEnum.CodingKeys.AutoCodable
        case someCase
        case anotherCase
        case id
        case name
// sourcery:end
    }
}

enum AssociatedValuesEnumNoCaseKey: AutoCodable, Equatable {
    case someCase(id: Int, name: String)
    case anotherCase
}
