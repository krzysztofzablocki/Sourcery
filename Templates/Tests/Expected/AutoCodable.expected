// Generated using Sourcery 0.12.0 â€” https://github.com/krzysztofzablocki/Sourcery
// DO NOT EDIT



extension AssociatedValuesEnum {

    internal init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        let enumCase = try container.decode(String.self, forKey: .enumCaseKey)
        switch CodingKeys(rawValue: enumCase) {
        case .someCase?:
            let id = try container.decode(Int.self, forKey: .id)
            let name = try container.decode(String.self, forKey: .name)
            self = .someCase(id: id, name: name)
        case .anotherCase?:
            self = .anotherCase
        default: throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: "Unknown enum case '\(enumCase)'"))
        }
    }

    internal func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        switch self {
        case let .someCase(id, name):
            try container.encode(CodingKeys.someCase.rawValue, forKey: .enumCaseKey)
            try container.encode(id, forKey: .id)
            try container.encode(name, forKey: .name)
        case .anotherCase:
            try container.encode(CodingKeys.anotherCase.rawValue, forKey: .enumCaseKey)
        }
    }

}


extension CustomCodingWithNotAllDefinedKeys {

    internal init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        value = try container.decode(Int.self, forKey: .value)
    }

    internal func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try container.encode(value, forKey: .value)
        encodeComputedValue(to: &container)
    }

}

extension CustomContainerCodable {

    public init(from decoder: Decoder) throws {
        let container = try CustomContainerCodable.decodingContainer(decoder)

        value = try container.decode(Int.self, forKey: .value)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encodingContainer(encoder)

        try container.encode(value, forKey: .value)
    }

}



extension CustomMethodsCodable {

    enum CodingKeys: String, CodingKey {
        case boolValue
        case intValue
        case optionalString
        case requiredString
        case requiredStringWithDefault
        case computedPropertyToEncode
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        boolValue = try CustomMethodsCodable.decodeBoolValue(from: decoder)
        intValue = CustomMethodsCodable.decodeIntValue(from: container) ?? CustomMethodsCodable.defaultIntValue
        optionalString = try container.decodeIfPresent(String.self, forKey: .optionalString)
        requiredString = try container.decode(String.self, forKey: .requiredString)
        requiredStringWithDefault = (try? container.decode(String.self, forKey: .requiredStringWithDefault)) ?? CustomMethodsCodable.defaultRequiredStringWithDefault
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try encodeBoolValue(to: encoder)
        encodeIntValue(to: &container)
        try container.encodeIfPresent(optionalString, forKey: .optionalString)
        try container.encode(requiredString, forKey: .requiredString)
        try container.encode(requiredStringWithDefault, forKey: .requiredStringWithDefault)
        encodeComputedPropertyToEncode(to: &container)
        try encodeAdditionalValues(to: encoder)
    }

}

extension SimpleEnum {

    enum CodingKeys: String, CodingKey {
        case someCase
        case anotherCase
    }

    internal init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()

        let enumCase = try container.decode(String.self)
        switch CodingKeys(rawValue: enumCase) {
        case .someCase?: self = .someCase
        case .anotherCase?: self = .anotherCase
        default: throw DecodingError.dataCorrupted(.init(codingPath: decoder.codingPath, debugDescription: "Unknown enum case '\(enumCase)'"))
        }
    }

    internal func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()

        switch self {
        case .someCase: try container.encode(CodingKeys.someCase.rawValue)
        case .anotherCase: try container.encode(CodingKeys.anotherCase.rawValue)
        }
    }

}

extension SkipDecodingWithDefaultValueOrComputedProperty {

    internal init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        value = try container.decode(Int.self, forKey: .value)
    }

    internal func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try container.encode(value, forKey: .value)
        try container.encode(computedValue, forKey: .computedValue)
    }

}

extension SkipEncodingKeys {

    enum CodingKeys: String, CodingKey {
        case value
        case skipValue
    }

    internal func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        try container.encode(value, forKey: .value)
    }

}
